003-graphql.md

```diff
@@ -0,0 +1,42 @@
+ - Because each application maintained its own caches, it was hard to maintain a consistent cache policy. (For example, one application might cache users for an hour while another didn't cache them at all).
+ - As our schema grew in each application, we became wary of adding new fields that would bloat responses. This slowed down development & made it harder to justify building shorter-term features or experiments.
+ - Finally, we had to do a lot of busy-work to [normalize our Redux state](https://redux.js.org/recipes/structuring-reducers/normalizing-state-shape) in Phoenix, which slowed down development of new features. (See [DoSomething/phoenix-next#767](https://github.com/DoSomething/phoenix-next/pull/767) and [DoSomething/phoenix-next#1123](https://github.com/DoSomething/phoenix-next/pull/1123)).
+
+## Proposal
+
+We built our GraphQL gateway (see [graphql.dosomething.org](https://graphql.dosomething.org/)) in November 2017. We have gradually added more resources (such as Contentful spaces and embed metadata) and applications (such as Gambit) over time.
+
+This gives us a single [universal schema](https://principledgraphql.com/integrity#1-one-graph) (so developers don't need to worry about which application holds which data). It lets us optimally [batch requests to backend services](https://www.apollographql.com/docs/graphql-tools/connectors/#dataloader-and-caching) and acts as a centralized "wrapper" around third-party APIs like [Contentful](https://www.contentful.com/).
+
+It also allows us to organically evolve our API over time (since we can add fields without bloating existing requests that don't need them, and [deprecate old fields](https://www.apollographql.com/docs/graphql-tools/generate-schema/#descriptions--deprecations) without breaking existing clients). We also get fantastic developer tools, like [GraphQL Playground](https://www.prisma.io/blog/introducing-graphql-playground-f1e0a018f05d), the [Apollo Chrome extension](https://chrome.google.com/webstore/detail/apollo-client-developer-t/jdkknkkbebbapilgoeccciglkfbmbnfm?hl=en-US) and [Visual Studio Code extension](https://marketplace.visualstudio.com/items?itemName=apollographql.vscode-apollo).
+
+We should continue to roll out GraphQL and reduce our reliance on custom data-fetching logic!
+
+## Drawbacks/Alternatives
+
+We've traditionally tried to [stick with "boring" technology](http://boringtechnology.club/), because we can follow established best practices & feel confident that we won't likely hit new edge cases. While GraphQL is used at scale [by many of our peers](https://www.apollographql.com/docs/intro/benefits/#ready-for-production), we know that we'll find more surprises than we might with REST & Redux.
+
+There's also an onboarding cost to using GraphQL. While it simplifies a lot of problems we've run into in the past, it also adds to the long list of technologies you need to be familiar with to be a productive DoSomething.org engineer.
```

Comment:
While I think there is a cost to onboarding to yet another technology, I think that graphql has enough baseline documentation, tutorials etc that make it fairly enjoyable to learn. (at least in my exp!)